core/drivers/sunxi_uart.c

```diff
@@ -0,0 +1,191 @@
+/* uart register defines */
+#define UART_REG_RBR	(0x00)
+#define UART_REG_THR	(0x00)
+#define UART_REG_DLL	(0x00)
+#define UART_REG_DLH	(0x04)
+#define UART_REG_IER	(0x04)
+#define UART_REG_IIR	(0x08)
+#define UART_REG_FCR	(0x08)
+#define UART_REG_LCR	(0x0c)
+#define UART_REG_MCR	(0x10)
+#define UART_REG_LSR	(0x14)
+#define UART_REG_MSR	(0x18)
+#define UART_REG_SCH	(0x1c)
+#define UART_REG_USR	(0x7c)
+#define UART_REG_TFL	(0x80)
+#define UART_REG_RFL	(0x84)
+#define UART_REG_HALT	(0xa4)
+
+/* uart status register bits */
+#define UART_REG_USR_BUSY BIT32(0)
+#define UART_REG_USR_TFNF BIT32(1)
+#define UART_REG_USR_TFE  BIT32(2)
+#define UART_REG_USR_RFNE BIT32(3)
+#define UART_REG_USR_RFF  BIT32(4)
+
+static vaddr_t chip_to_base(struct serial_chip *chip)
+{
+	struct sunxi_uart_data *pd =
+		container_of(chip, struct sunxi_uart_data, chip);
+
+	return io_pa_or_va(&pd->base);
+}
+
+static void sunxi_uart_flush(struct serial_chip *chip)
+{
+	vaddr_t base = chip_to_base(chip);
+
+	while (read32(base + UART_REG_TFL)) {
+		/* waiting transmit fifo empty */
+		;
+	}
+}
+
+static bool sunxi_uart_have_rx_data(struct serial_chip *chip)
+{
+	vaddr_t base = chip_to_base(chip);
+
+	return read32(base + UART_REG_RFL);
+}
+
+static int sunxi_uart_getchar(struct serial_chip *chip)
+{
+	vaddr_t base = chip_to_base(chip);
+
+	while (!sunxi_uart_have_rx_data(chip)) {
+		/* transmit fifo is empty, waiting again. */
+		;
+	}
+	return read32(base + UART_REG_RBR) & 0xff;
+}
+
+static void sunxi_uart_putc(struct serial_chip *chip, int ch)
+{
+	vaddr_t base = chip_to_base(chip);
+
+	while (!(read32(base + UART_REG_USR) & UART_REG_USR_TFNF)) {
+		/* transmit fifo is full, waiting again. */
+		;
+	}
+
+	/* write out charset to transmit fifo */
+	write8(ch, base + UART_REG_THR);
+}
+
+static const struct serial_ops sunxi_uart_ops = {
+	.flush = sunxi_uart_flush,
+	.getchar = sunxi_uart_getchar,
+	.have_rx_data = sunxi_uart_have_rx_data,
+	.putc = sunxi_uart_putc,
+};
+KEEP_PAGER(sunxi_uart_ops);
+
+void sunxi_uart_init(struct sunxi_uart_data *pd, paddr_t base)
+{
+	pd->base.pa = base;
+	pd->chip.ops = &sunxi_uart_ops;
+
+	/*
+	 * Do nothing, debug uart(uart0) share with normal world,
+	 * everything for uart0 is ready now.
+	 */
+}
+
+#ifdef CFG_DT
+
+static struct serial_chip *sunxi_dev_alloc(void)
+{
+	struct sunxi_uart_data *pd = malloc(sizeof(*pd));
+
+	if (!pd)
+		return NULL;
+	return &pd->chip;
+}
+
+static int sunxi_dev_init(struct serial_chip *chip, const void *fdt, int offs,
+			  const char *parms)
+{
+	struct sunxi_uart_data *pd =
+		container_of(chip, struct sunxi_uart_data, chip);
+	vaddr_t vbase;
+	paddr_t pbase;
+	size_t size;
+
+	if (parms && parms[0])
+		IMSG("sunxi: device parameters ignored (%s)", parms);
+
+	if (dt_map_dev(fdt, offs, &vbase, &size) < 0)
+		return -1;
+
+	if (size != SUNXI_UART_REG_SIZE) {
+		EMSG("sunxi: unexpected register size: %zx", size);
+		return -1;
+	}
+
+	pbase = virt_to_phys((void *)vbase);
+	IMSG("sunxi: base: %lx", pbase);
```

Comment:
`s/sunxi/sunxi_uart/` and demote to `DMSG()` (or remove altogether? is this message useful?)